docs/src/references/mappers/gen-mapper.md

```diff
@@ -0,0 +1,121 @@
+  - `MQTT sub| filter-1 | filter-2 | ... | filter-n | MQTT pub`
+- A pipeline can combine builtin and user-provided filters.
+- The user can configure all the transformations used by a mapper,
+  editing MQTT sources, pipelines, filters and MQTT sinks.
+- By contrast with the current implementation, where the translation of measurements from %%te%% JSON to Cumulocity JSON
+  is fully hard-coded, with the generic mapper a user can re-use the core of this transformation while adding customized steps:
+  - consuming measurement from a non-standard topic
+  - filtering out part of the measurements
+  - normalizing units
+  - adding units read from some config
+  - producing transformed measurements on a non-standard topic.
+
+## POC reference
+
+- The generic mapper loads pipeline and filters stored in `/etc/tedge/gen-mapper/`.
+- A pipeline is defined by a TOML file with `.toml` extension.
+- A filter is defined by a Javascript file with `.js` extension.
+- The definition of pipeline must provide a list of MQTT topics to subscribe to.
+  - The pipeline will be feed with all the messages received on these topics.
+- A pipeline definition also provides a list of stages.
+  - Each stage is built from a javascript and is possibly given a config (arbitrary json that will be passed to the script)
+  - Each stage can also subscribe to a list of MQTT topics (which messages will be passed to the script to update its config)
+
+```toml
+input_topics = ["te/+/+/+/+/m/+"]
+
+stages = [
+    { filter = "add_timestamp.js" },
+    { filter = "drop_stragglers.js", config = { max_delay = 60 } },
+    { filter = "te_to_c8y.js", meta_topics = ["te/+/+/+/+/m/+/meta"] }
+]
+```
+
+- A filter has to export at least one `process` function.
+  - This function is called for each message to be transformed
+  - The arguments passed to the function are:
+    - The current time as `{ seconds: u64, nanoseconds: u32 }` 
+    - The message `{ topic: string, payload: string }`
+    - The config as read from the pipeline config or updated by the script
+  - The function is expected to return zero, one or many transformed messages `[{ topic: string, payload: string }]`
+  - An exception can be thrown if the input message cannot be transformed.
+- A filter can also export an `update_config` function
+  - This function is called on each message received on the `meta_topics` as defined in the config.
+  - The arguments are:
+    - The message to be interpreted as a config update `{ topic: string, payload: string }`
+    - The current config
+   - The returned value (an arbitrary JSON value) is then used as the new config for the filter.
+- A filter can also export a `tick` function
+  - This function is called at a regular pace with the current time and config.
+  - The filter can then return zero, one or many transformed messages
+  - By sharing an internal state between the `process` and `tick` functions,
```

Comment:
I think I get the purpose of `tick` in isolation, as a way of aggregating data points. But, I'm failing to understand the use-case for combining both `process` and `tick` together. I hope it'll be clearer as more examples are added.

Reply:
See https://github.com/thin-edge/thin-edge.io/pull/3650/commits/b5a9c83c8495c266eb51a34b3832bfc41a5c4bd9